File: src/quantized/routes/thin_film.py
"""Thin-film calculator routes. Wraps ``calc.thin_film`` (pure formulas):
deposition / sputter rate, diffusion length, implant dose + peak concentration,
Kiessig thickness, multilayer thermal conductivity, projected range, Stoney
stress, thermal-mismatch strain. Validate -> call the pure fn -> serialize.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from quantized.calc import thin_film

router = APIRouter(prefix="/api/thin-film", tags=["thin-film"])


class DepositionRateRequest(BaseModel):
    thickness: float  # Å
    time: float  # s


class DiffusionLengthRequest(BaseModel):
    d: float  # cm^2/s
    t: float  # s


class DoseFromCurrentRequest(BaseModel):
    current: float  # A
    time: float  # s
    area: float  # cm^2


class DoseToConcentrationRequest(BaseModel):
    dose: float  # ions/cm^2
    rp: float  # nm
    delta_rp: float  # nm


class KiessigRequest(BaseModel):
    delta_q: float  # Å^-1
    sld: float | None = None  # Å^-2
    qc: float | None = None  # Å^-1


class MultilayerThermalRequest(BaseModel):
    thicknesses: list[float]  # nm
    kappas: list[float]  # W/m/K


class ProjectedRangeRequest(BaseModel):
    ion: str
    target: str
    energy: float  # keV


class SputterRateRequest(BaseModel):
    y: float  # atoms/ion
    j: float  # mA/cm^2
    rho: float  # g/cm^3
    m: float  # g/mol


class StoneyStressRequest(BaseModel):
    es: float  # Pa
    nus: float
    ts: float  # m
    tf: float  # m
    r: float  # m


class ThermalMismatchRequest(BaseModel):
    alpha_film: float  # 1/K
    alpha_sub: float  # 1/K
    delta_t: float  # K
    e: float | None = None  # Pa
    nu: float = 0.3


def _call(fn: Callable[..., dict[str, Any]], *args: Any, **kwargs: Any) -> dict[str, Any]:
    try:
        return fn(*args, **kwargs)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/deposition-rate")
def deposition_rate(req: DepositionRateRequest) -> dict[str, Any]:
    """r = thickness/time (Å/s, nm/min)."""
    return _call(thin_film.deposition_rate, req.thickness, req.time)


@router.post("/diffusion-length")
def diffusion_length(req: DiffusionLengthRequest) -> dict[str, Any]:
    """L = √(D·t) (cm/nm/µm)."""
    return _call(thin_film.diffusion_length_thermal, req.d, req.t)


@router.post("/dose-from-current")
def dose_from_current(req: DoseFromCurrentRequest) -> dict[str, Any]:
    """Φ = I·t/(q·A) (ions/cm²)."""
    return _call(thin_film.dose_from_current, req.current, req.time, req.area)


@router.post("/dose-to-concentration")
def dose_to_concentration(req: DoseToConcentrationRequest) -> dict[str, Any]:
    """C_peak = dose/(√(2π)·ΔRp) (atoms/cm³)."""
    return _call(thin_film.dose_to_concentration, req.dose, req.rp, req.delta_rp)


@router.post("/kiessig-thickness")
def kiessig_thickness(req: KiessigRequest) -> dict[str, Any]:
    """t = 2π/ΔQ (refraction-corrected when SLD/Qc supplied)."""
    return _call(thin_film.kiessig_thickness, req.delta_q, sld=req.sld, qc=req.qc)


@router.post("/multilayer-thermal")
def multilayer_thermal(req: MultilayerThermalRequest) -> dict[str, Any]:
    """Series / parallel effective thermal conductivity (W/m/K)."""
    return _call(thin_film.multilayer_thermal_conductivity, req.thicknesses, req.kappas)


@router.post("/projected-range")
def projected_range(req: ProjectedRangeRequest) -> dict[str, Any]:
    """LSS projected range Rp + straggle ΔRp (nm)."""
    return _call(thin_film.projected_range, req.ion, req.target, req.energy)


@router.post("/sputter-rate")
def sputter_rate(req: SputterRateRequest) -> dict[str, Any]:
    """Sputter erosion rate (nm/s, nm/min)."""
    return _call(thin_film.sputter_rate, req.y, req.j, req.rho, req.m)


@router.post("/stoney-stress")
def stoney_stress(req: StoneyStressRequest) -> dict[str, Any]:
    """σ = E_s·t_s²/(6(1−ν_s)·t_f·R) (Pa)."""
    return _call(thin_film.stoney_stress, req.es, req.nus, req.ts, req.tf, req.r)


@router.post("/thermal-mismatch")
def thermal_mismatch(req: ThermalMismatchRequest) -> dict[str, Any]:
    """ε = (α_f−α_s)·ΔT, σ = E·ε/(1−ν) (MPa)."""
    return _call(
        thin_film.thermal_mismatch_strain,
        req.alpha_film,
        req.alpha_sub,
        req.delta_t,
        e=req.e,
        nu=req.nu,
    )

File: frontend/src/lib/api/thinFilm.ts
// /api/thin-film/* wrappers. `thinFilmSauerbrey` was extracted first (the
// lib/api.ts pin had zero headroom); the original ten joined it in the R8
// bundle-diet pass (2026-08-23, see api/reference.ts's header for why) —
// they were still defined directly in lib/api.ts, so useApp.ts's unrelated
// eager imports from that SAME file were dragging this lazy-only
// (thinfilm/ cards only) calculator set into the eager bundle purely by
// file co-location. NOT re-exported by lib/api.ts (zero headroom there);
// the thinfilm/ cards import everything directly from this path.

import { postJSON } from "./http";

/** Deposition rate = thickness/time. */
export function thinFilmDepositionRate(
  thickness: number,
  time: number,
): Promise<{ rate: number; rate_nm_per_min: number }> {
  return postJSON("/api/thin-film/deposition-rate", { thickness, time });
}

/** Sputter deposition rate from yield, current density, density, molar mass. */
export function thinFilmSputterRate(
  y: number,
  j: number,
  rho: number,
  m: number,
): Promise<{ rate: number; rate_nm_per_min: number }> {
  return postJSON("/api/thin-film/sputter-rate", { y, j, rho, m });
}

/** Thermal diffusion length L = √(D·t). */
export function thinFilmDiffusionLength(
  d: number,
  t: number,
): Promise<{ L: number; L_nm: number; L_um: number }> {
  return postJSON("/api/thin-film/diffusion-length", { d, t });
}

/** Implant dose = I·t/(q·A). */
export function thinFilmDoseFromCurrent(
  current: number,
  time: number,
  area: number,
): Promise<{ dose: number }> {
  return postJSON("/api/thin-film/dose-from-current", { current, time, area });
}

/** Peak concentration from dose + projected-range straggle. */
export function thinFilmDoseToConcentration(
  dose: number,
  rp: number,
  deltaRp: number,
): Promise<{ Cpeak: number }> {
  return postJSON("/api/thin-film/dose-to-concentration", { dose, rp, delta_rp: deltaRp });
}

/** Kiessig fringe thickness from Δq (with refraction correction). */
export function thinFilmKiessig(
  deltaQ: number,
  sld?: number,
  qc?: number,
): Promise<{ thickness: number; thickness_nm: number; Qc: number; thickness_raw: number }> {
  return postJSON("/api/thin-film/kiessig-thickness", { delta_q: deltaQ, sld, qc });
}

/** Series + parallel thermal conductivity of a multilayer stack. */
export function thinFilmMultilayerThermal(
  thicknesses: number[],
  kappas: number[],
): Promise<{ k_series: number; k_parallel: number; total_thickness: number; n_layers: number }> {
  return postJSON("/api/thin-film/multilayer-thermal", { thicknesses, kappas });
}

/** LSS projected range R_p + straggle ΔR_p. */
export function thinFilmProjectedRange(
  ion: string,
  target: string,
  energy: number,
): Promise<{ Rp: number; deltaRp: number; warning: string }> {
  return postJSON("/api/thin-film/projected-range", { ion, target, energy });
}

/** Stoney film stress from substrate curvature. */
export function thinFilmStoneyStress(
  es: number,
  nus: number,
  ts: number,
  tf: number,
  r: number,
): Promise<{ stress: number; stress_MPa: number; stress_GPa: number }> {
  return postJSON("/api/thin-film/stoney-stress", { es, nus, ts, tf, r });
}

/** Thermal-mismatch strain/stress between film and substrate. */
export function thinFilmThermalMismatch(
  alphaFilm: number,
  alphaSub: number,
  deltaT: number,
  e?: number,
  nu?: number,
): Promise<{ strain: number; stress_MPa: number; description: string }> {
  return postJSON("/api/thin-film/thermal-mismatch", {
    alpha_film: alphaFilm,
    alpha_sub: alphaSub,
    delta_t: deltaT,
    e,
    nu,
  });
}

/** QCM areal mass (+ total mass/thickness when area/density given) from a
 *  Sauerbrey frequency shift. `delta_f` is negative for added mass. */
export function thinFilmSauerbrey(
  deltaF: number,
  f0: number,
  area?: number,
  density?: number,
): Promise<{
  areal_mass: number;
  areal_mass_ng_cm2: number;
  Cf: number;
  Cf_hz_cm2_ug: number;
  delta_m?: number;
  delta_m_ng?: number;
  thickness?: number;
  thickness_nm?: number;
}> {
  return postJSON("/api/thin-film/sauerbrey", { delta_f: deltaF, f0, area, density });
}

/** Scherrer crystallite size from FWHM (degrees 2-theta), wavelength (Angstrom),
 *  and peak position (degrees 2-theta), using DiraCulator's fixed K=0.9. */
export function thinFilmScherrer(
  fwhmDeg: number,
  wavelength: number,
  twoThetaDeg: number,
): Promise<{ D: number; D_nm: number; K: number }> {
  return postJSON("/api/thin-film/scherrer", {
    fwhm_deg: fwhmDeg,
    wavelength,
    two_theta_deg: twoThetaDeg,
  });
}
